src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1036,322 @@ begin
+  { simp }
+end
+
+/-- Given a bilinear form `B` and some `x` such that `B x x ≠ 0`, the span of the singleton of `x`
+  is complement to its orthogonal complement. -/
+lemma is_compl_singleton {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  is_compl (submodule.span K ({x} : set V))
+    (B.orthogonal (submodule.span K ({x} : set V))) :=
+{ inf_le_bot := eq_bot_iff.1 $ span_inf_ortho_eq_bot hx,
+  top_le_sup := eq_top_iff.1 $ span_sup_ortho_eq_top hB hx }
+
+/-- The natural isomorphism between a singleton and the quotient by its orthogonal complement. -/
+noncomputable def quotient_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.quotient_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+/-- The natural isomorphism from the product between a singleton and its orthogonal component
+  and the whole space. -/
+noncomputable def prod_equiv_of_ortho_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :=
+submodule.prod_equiv_of_is_compl _ _ (is_compl_singleton hB hx)
+
+lemma restrict_ortho_singleton_nondegenerate (B : bilin_form K V) (hB₁ : nondegenerate B)
+  (hB₂ : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  nondegenerate $ B.restrict $ B.orthogonal (submodule.span K ({x} : set V)) :=
+begin
+  refine λ m hm, submodule.coe_eq_zero.1 (hB₁ m.1 (λ n, _)),
+  have : n ∈ submodule.span K ({x} : set V) ⊔
+    B.orthogonal (submodule.span K ({x} : set V)) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw restrict at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+/- Let V be a finite dimensional vector space over the field K with the
+  nondegenerate bilinear form B. Then for all m ∈ M, f_m : M → R₁ : n ↦ B m n is
+  a linear functional in the dual space.
+
+  Furthermore, the map, φ : M → M* : m ↦ f_m is an isomorphism. -/
+
+/-- Given a bilinear form `B` on the modual `M`, `to_dual' B` is the linear map
+  from `M` to its dual such that `to_dual B m` is the functional `λ x, B m x`. -/
+def to_dual' (B : bilin_form R₃ M₃) : M₃ →ₗ[R₃] module.dual R₃ M₃ :=
+B.to_lin
+
+@[simp] lemma to_dual'_def {B : bilin_form R₃ M₃} {m n : M₃} :
+  B.to_dual' m n = B m n := rfl
+
+lemma to_dual'_injective {B : bilin_form R₃ M₃} (hB : B.nondegenerate) :
+  function.injective (to_dual' B) :=
+to_lin_injective hB
+
+open finite_dimensional
+
+variable [finite_dimensional K V]
+
+-- In order to show that `to_dual` is a surjective map we used the fact that
+-- the dimensions of a vector space equal to the dimensions of its dual.
+-- So rather than working with modules over rings, we work with vector spaces
+lemma to_dual'_bijective (B : bilin_form K V) (hB : B.nondegenerate) :
+  function.bijective (to_dual' B) :=
```

Comment:
Instead of this lemma, can you extract a general construction (if it doesn't already exist) of a `linear_equiv` from a `linear_map` between two vector spaces whose `findim` is the same?  Not sure where it should live, maybe `linear_algebra.finite_dimensional`.

Reply:
I've added `linear_equiv_of_ker_eq_bot` for this purpose